### ISSUE TITLE
Can I use two GPUs for VLLM?

### ISSUE THREAD

**Issue by LLMforScience** [huggingface/open-r1#471]
Thank you for your work.

I understand that you are using 1 GPU for VLLM inference and the other 7 GPUs for training.

Can I use two GPUs for VLLM inference? I noticed that using 1 GPU results in an out-of-memory (OOM) error when I set the max_completion_length to a large value.

**Comment by samma1570** [comment]
same problem

**Comment by tastelikefeet** [comment]
you can try [SWIFT](https://github.com/modelscope/ms-swift/tree/main/examples/train/grpo)

**Comment by rachelsyr** [comment]
> you can try [SWIFT](https://github.com/modelscope/ms-swift/tree/main/examples/train/grpo) based on TRL and supports multi vllms and tensor-parallel

Now I have 4 GPUs. I set num_infer_workers=2 to use 2 GPUs for vLLM inference, but I get an error: 'world_size (2) is not equal to tensor_model_parallel_size (1) x pipeline_model_parallel_size (1)'. Then I set tensor_parallel_size=2, but I get another error: 'tensor_parallel_size>1 only supports num_infer_workers==your available device count.' What should I do to use multi-GPU vLLM

**Comment by tastelikefeet** [comment]
> > you can try [SWIFT](https://github.com/modelscope/ms-swift/tree/main/examples/train/grpo) based on TRL and supports multi vllms and tensor-parallel
> 
> Now I have 4 GPUs. I set num_infer_workers=2 to use 2 GPUs for vLLM inference, but I get an error: 'world_size (2) is not equal to tensor_model_parallel_size (1) x pipeline_model_parallel_size (1)'. Then I set tensor_parallel_size=2, but I get another error: 'tensor_parallel_size>1 only supports num_infer_workers==your available device count.' What should I do to use multi-GPU vLLM

I met this issue when I used an old vllm version. Can you share your command line and vllm version here please?
Now SWIFT supports two device placement:
1. Part of GPUs use for training, another part of GPUs for rollout https://github.com/modelscope/ms-swift/blob/main/examples/train/grpo/full_vllm.sh
2. Put Model and vLLM together, this placement supports vllm tensor_parallel https://github.com/modelscope/ms-swift/blob/main/examples/train/grpo/train_72b_4gpu.sh

**Comment by rachelsyr** [comment]
> > > you can try [SWIFT](https://github.com/modelscope/ms-swift/tree/main/examples/train/grpo) based on TRL and supports multi vllms and tensor-parallel
> > 
> > 
> > Now I have 4 GPUs. I set num_infer_workers=2 to use 2 GPUs for vLLM inference, but I get an error: 'world_size (2) is not equal to tensor_model_parallel_size (1) x pipeline_model_parallel_size (1)'. Then I set tensor_parallel_size=2, but I get another error: 'tensor_parallel_size>1 only supports num_infer_workers==your available device count.' What should I do to use multi-GPU vLLM
> 
> I met this issue when I used an old vllm version. Can you share your command line and vllm version here please? Now SWIFT supports two device placement:
> 
> 1. Part of GPUs use for training, another part of GPUs for rollout https://github.com/modelscope/ms-swift/blob/main/examples/train/grpo/full_vllm.sh
> 2. Put Model and vLLM together, this placement supports vllm tensor_parallel https://github.com/modelscope/ms-swift/blob/main/examples/train/grpo/train_72b_4gpu.sh

My vLLM version is 0.7.2, and the script I'm using is modified based on full_vllm.sh:
"export CUDA_VISIBLE_DEVICES=0,1,2,3
export NPROC_PER_NODE=2

swift rlhf \
    --rlhf_type grpo \
    --model /data/home/models/Qwen2.5-7B-Instruct \
    --reward_funcs accuracy format \
    --use_vllm true \
    --num_infer_workers 2 \
    --vllm_gpu_memory_utilization 0.8 \
    --vllm_max_model_len 8192 \
    --train_type full \
    --torch_dtype bfloat16 \
    --dataset /data/home/datasets/NuminaMath-TIR#5000 \
    --max_completion_length 1024 \
    --num_train_epochs 1 \
    --per_device_train_batch_size 4 \
    --per_device_eval_batch_size 4 \
    --learning_rate 1e-6 \
    --gradient_accumulation_steps 2 \
    --eval_steps 200 \
    --save_steps 200 \
    --save_total_limit 2 \
    --logging_steps 5 \
    --max_length 4096 \
    --output_dir output \
    --warmup_ratio 0.05 \
    --dataloader_num_workers 4 \
    --dataset_num_proc 4 \
    --num_generations 4 \
    --temperature 0.9 \
    --system examples/train/grpo/prompt.txt \
    --deepspeed zero2 \
    --log_completions true \
    --async_generate true"

**Comment by rachelsyr** [comment]
> > > you can try [SWIFT](https://github.com/modelscope/ms-swift/tree/main/examples/train/grpo) based on TRL and supports multi vllms and tensor-parallel
> > 
> > 
> > Now I have 4 GPUs. I set num_infer_workers=2 to use 2 GPUs for vLLM inference, but I get an error: 'world_size (2) is not equal to tensor_model_parallel_size (1) x pipeline_model_parallel_size (1)'. Then I set tensor_parallel_size=2, but I get another error: 'tensor_parallel_size>1 only supports num_infer_workers==your available device count.' What should I do to use multi-GPU vLLM
> 
> I met this issue when I used an old vllm version. Can you share your command line and vllm version here please? Now SWIFT supports two device placement:
> 
> 1. Part of GPUs use for training, another part of GPUs for rollout https://github.com/modelscope/ms-swift/blob/main/examples/train/grpo/full_vllm.sh
> 2. Put Model and vLLM together, this placement supports vllm tensor_parallel https://github.com/modelscope/ms-swift/blob/main/examples/train/grpo/train_72b_4gpu.sh

Thank you for your suggestion. I was able to solve the full vllm error issue simply by upgrading vllm from 0.7.2 to 0.7.3 (which is strange, since 0.7.2 is the version required by openr1 and not a particularly old version). In addition, the tensor_parallel approach is now working normally as well. Do you know which of these two configuration methods is more efficient for training?
